File: algo/datastruct/tree/binary-search-tree/src/BSTree.hs
{-
 BSTree.hs
 Copyright (C) 2010 Liu Xinyu (liuxinyu95@gmail.com)
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-}

module BSTree where

import Test.QuickCheck
import Data.List (sort)

data Tree a = Empty 
            | Node (Tree a) a (Tree a) deriving (Show)

-- Helper functions for tree
leaf::a -> Tree a
leaf a = Node Empty a Empty

left::Tree a -> Tree a
left (Node l _ _) = l
left _ = Empty

right::Tree a -> Tree a
right (Node _ _ r) = r
right _ = Empty

key::Tree a -> a
key (Node _ k _) = k

isEmpty::Tree a -> Bool
isEmpty Empty = True
isEmpty _ = False

-- in-order tree traverse
mapT::(a->b) -> Tree a -> Tree b
mapT _ Empty = Empty
mapT f (Node l x r)= Node (mapT f l) (f x) (mapT f r)

-- Search in tree
search::(Ord a)=> Tree a -> a -> Tree a
search Empty _ = Empty
search t@(Node l k r) x | k == x = t
                        | x < k = search l x
                        | otherwise = search r x

-- Tree Min
mint::Tree a -> Tree a
mint t = if isEmpty (left t) then t else mint $ left t

-- Tree Max
maxt::Tree a -> Tree a
maxt t = if isEmpty (right t) then t else maxt $ right t

-- Below functions are low efficiency compare to their imperactive implementation

-- Find parent of a value in tree
-- The performance of parent is low: O(log n)
parent::(Ord a)=>Tree a -> a -> Tree a
parent Empty _ = Empty
parent t x | x < (key t) = if isParent (left t) x then t else parent (left t) x
           | x > (key t) = if isParent (right t) x then t else parent (right t) x
           | otherwise = Empty where
               isParent Empty _ = False
               isParent t x = (key t)==x

-- Find an ancestor of a value wich match a certain rule
-- Performance is very low.
findAncestor::(Ord a)=>Tree a -> a ->(a->a->Bool)-> Tree a
findAncestor t x f = checkParent t (parent t x) x f where
    checkParent _ Empty _ _ = Empty
    checkParent t p x f = if f (key p) x then p
                          else checkParent t (parent t (key p)) x f


-- successor
-- Performance is very low.
succt::(Ord a)=>Tree a -> a -> Tree a
succt Empty _ = Empty
succt t x = if not $ isEmpty rightNode
            then mint rightNode
            else findAncestor t x (>) where
                rightNode = right (search t x)

-- predecessor
-- Performance is very low
predt::(Ord a)=>Tree a -> a -> Tree a
predt Empty _ = Empty
predt t x = if not $ isEmpty leftNode
            then maxt leftNode
            else findAncestor t x (<) where
                leftNode = left (search t x)

-- End of low efficiency functions
              
-- Insert an element into a tree
insert::(Ord a) => Tree a -> a -> Tree a
insert Empty k = Node Empty k Empty
insert (Node l x r) k | k < x = Node (insert l k) x r
                      | otherwise = Node l x (insert r k)

-- Delete an element from a tree
-- The algorithm described in CLRS is not used here, I used the algorithm
-- which is mentioned in Annotated STL, P 235 (by Hou Jie)
--   if x has only one child: just splice x out
--   if x has two children: use min(right) to replce x
delete::(Ord a)=> Tree a -> a -> Tree a
delete Empty _ = Empty
delete (Node l k r) x | x < k = (Node (delete l x) k r)
                      | x > k = (Node l k (delete r x))
                      -- x == k
                      | isEmpty l = r
                      | isEmpty r = l
                      | otherwise = (Node l k' (delete r k')) where k' = key $ mint r

-- Helper to build a binary search tree from a list
fromList::(Ord a)=>[a] -> Tree a
fromList = foldl insert Empty

toList::(Ord a)=>Tree a -> [a]
toList Empty = []
toList (Node l x r) = toList l ++ [x] ++ toList r

-- test
prop_build :: (Ord a)=>[a] -> Bool
prop_build xs = sort xs == (toList $ fromList xs)

-- test data
t1 = leaf 4
t2 = fromList [15, 6, 18, 3, 7, 17, 20, 2, 4, 13, 9]

-- test tree creation
testBuildTree = "\ntest create empty:\t"++ show (Empty::Tree Int)++
                "\ntest create leaf:\t" ++ show t1 ++ 
                "\ntest create from list:\t" ++ show t2

-- test tree walk
testTreeWalk = "\ntest tree in-order walk by apply (-):\t"++show (mapT negate t2)

-- test min/max for tree
testMinMax = "\ntest min empty:\t" ++ (show $ mint (Empty::Tree Int)) ++
             "\ntest min leaf:\t" ++  show (mint t1) ++
             "\ntest min tree:\t" ++ show (mint t2) ++ 
             "\ntest max empty:\t" ++ (show $ maxt (Empty::Tree Int)) ++
             "\ntest max leaf:\t" ++ show (maxt t1) ++
             "\ntest max tree:\t" ++ show (maxt t2)

-- test search in tree
testSearch = "\ntest search empty tree:\t"++ show (search (Empty::Tree Int) 3) ++
             "\ntest search in leaf:\t"++ show (search t1 4)++
             "\ntest search non exist value in leaf:\t" ++ show (search t1 5)++
             "\ntest search non exist node in tree:\t"++ show (search t2 5)++
             "\ntest search a node in tree:\t"++ show (search t2 18)

-- test succ/pred
testSuccPred = "\ntest succ of 7:\t"++ show (key (succt t2 7)) ++
               "\ntest parent of 7:\t"++ show (key (parent t2 7))++
               "\ntest parent of 13:\t"++ show (key (parent t2 13))++
               "\ntest bigger ancestor of 7:\t"++ show (key (findAncestor t2 7 (>)))++
               "\ntest bigger ancestor of 13:\t"++ show (key (findAncestor t2 13 (>)))++
               "\ntest succ of 13:\t"++ show (key (succt t2 13)) ++
               "\ntest pred of 6:\t"++ show (key (predt t2 6))++
               "\ntest pred of 7:\t"++ show (key (predt t2 7))

testDel = "\ntest del 17:\t"++ show (delete t2 17)++
          "\ntest del 7:\t"++ show (delete t2 7)++
          "\ntest del 6:\t" ++ show (delete t2 6)++
          "\ntest del 15:\t"++ show (delete t2 15)++
          "\ntest del non-exist:\t" ++ show (delete t2 5)

oldTest = do
    putStrLn testBuildTree
    putStrLn testMinMax
    putStrLn testSearch
    putStrLn testTreeWalk
    putStrLn testSuccPred
    putStrLn testDel

--newTest = do
--    test prop_build

File: algo/datastruct/tree/binary-search-tree/src/bstree.py
#!/usr/bin/python

# bstree.py
# Copyright (C) 2010 Liu Xinyu (liuxinyu95@gmail.com)
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
# 

# I tested that python doesn't support tail recursion.
# as below:
#
# Program A
#
# def fsum(n):
#     if(n==0):
#         return 0
#     else:
#         return n+fsum(n-1)
#
# Program B
#
# def fact(n, acc = 1):
#        if n == 0:
#                return acc
#        else:
#                return fact(n-1, acc*n)  
# 
# if we call fsum(1000) or fact(1000), the error is:
# RuntimeError: maximum recursion depth exceeded

import sys
import string

# Definition

class Node:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.parent = None #optional, it's helpful for succ/pred

def in_order_walk(t, f):
    if(t!=None):
        in_order_walk(t.left, f)
        f(t.key)
        in_order_walk(t.right, f)

# Querying

def tree_search(t, x):
    while(t!=None and t.key != x):
        if(x < t.key): t = t.left
        else: t = t.right
    return t

def tree_min(t):
    while(t!=None and t.left != None):
        t = t.left
    return t

def tree_max(t):
    while(t!=None and t.right != None):
        t = t.right
    return t

def succ(x):
    if x is None: return None
    if x.right is not None: return tree_min(x.right)
    p = x.parent
    while p is not None and p.left != x:
        x = p
        p = p.parent
    return p

def pred(x):
    if x is None: return None
    if x.left is not None: return tree_max(x.left)
    p = x.parent
    while p is not None and p.right != x:
        x = p
        p = p.parent
    return p

# Insertion and deletion

def tree_insert(t, key):
    root = t
    x = Node(key)
    parent = None
    while(t):
        parent = t
        if(key < t.key):
            t = t.left
        else:
            t = t.right
    x.parent = parent
    if(parent == None): #tree is empty
        return x
    elif(key < parent.key):
        parent.left = x
    else:
        parent.right = x
    return root

def remove_node(x):
    if (x is None): return
    x.parent = x.left = x.right = None

def tree_delete(t, x):
    if x is None: 
        return t
    [root, old_x, parent] = [t, x, x.parent]
    if x.left is None:
        x = x.right
    elif x.right is None:
        x = x.left
    else:
        y = tree_min(x.right)
        x.key = y.key
        if y.parent != x:
            y.parent.left = y.right
        else:
            x.right = y.right
        remove_node(y)
        return root
    if x is not None:
        x.parent = parent
    if parent is None:
        root = x
    else:
        if parent.left == old_x: 
            parent.left = x
        else: 
            parent.right = x
    remove_node(old_x)
    return root

# Helper functions

def list_to_tree(l):
    tree = None
    for x in l:
        tree = tree_insert(tree, x)
    return tree

def clone_tree(t, parent = None):
    n = None
    if(t != None):
        n = Node(t.key)
        n.left = clone_tree(t.left, n)
        n.right = clone_tree(t.right, n)
        n.parent = parent
    return n

def tree_to_str(t):
    if(t!=None):
        return "("+tree_to_str(t.left)+"), " + str(t.key) + ", (" + tree_to_str(t.right)+")"
    else:
        return "empty"

def my_print(x):
    print x

class Test:
    def __init__(self):
        self.tree = list_to_tree([15, 6, 18, 3, 7, 17, 20, 2, 4, 13, 9])
        print tree_to_str(self.tree)

    def __assert(self, msg, x, y):
        if(x == y): msg = msg + "OK."
        else: msg = msg + str(x) + "!=" + str(y) + "Fail."
        print msg
    
    def run(self):
        self.test_in_order_walk()
        self.test_search()
        self.test_min_max()
        self.test_succ_pred()
        self.test_del()

    def test_in_order_walk(self):
        print "test in order walk with my_print: "
        in_order_walk(self.tree, my_print)

    def test_search(self):
        self.__assert("search empty: ", tree_search(None, 3), None)
        print tree_to_str(tree_search(self.tree, 18))
        self.__assert("search non-exist: ", tree_search(self.tree, 5), None)

    def test_min_max(self):
        self.__assert("min(empty): ", tree_min(None), None)
        self.__assert("min(tree): ", tree_min(self.tree).key, 2)
        self.__assert("max(empty): ", tree_max(None), None)
        self.__assert("max(tree): ", tree_max(self.tree).key, 20)

    def test_succ_pred(self):
        self.__assert("succ of 7: ", succ(tree_search(self.tree, 7)).key, 9)
        self.__assert("succ of 13: ", succ(tree_search(self.tree, 13)).key, 15)
        self.__assert("succ of 20: ", succ(tree_search(self.tree, 20)), None)
        self.__assert("pred of 6: ", pred(tree_search(self.tree, 6)).key, 4)
        self.__assert("pred of 7: ", pred(tree_search(self.tree, 7)).key, 6)
        self.__assert("pred of 2: ", pred(tree_search(self.tree, 2)), None)

    def __test_del_n(self, n):
        t = clone_tree(self.tree)
        t = tree_delete(t, tree_search(t, n))
        print "del ", n, ": ", tree_to_str(t)
        self.__assert("search after del: ", tree_search(t, n), None)

    def test_del(self):
        self.__test_del_n(17)
        self.__test_del_n(7)
        self.__test_del_n(6)
        self.__test_del_n(15)
        self.__test_del_n(1) #del a non-exist value

if __name__ == "__main__":
    Test().run()
